File: 0123_Best_Time_to_Buy_and_Sell_Stock_III/solution.java
package cn.huang.leetcode;
/*
123. Best Time to Buy and Sell Stock III

Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete at most two transactions.

Note: You may not engage in multiple transactions at the same time (i.e., you must sell the stock before you buy again).

Example 1:

Input: [3,3,5,0,0,3,1,4]
Output: 6
Explanation: Buy on day 4 (price = 0) and sell on day 6 (price = 3), profit = 3-0 = 3.
             Then buy on day 7 (price = 1) and sell on day 8 (price = 4), profit = 4-1 = 3.

Example 2:

Input: [1,2,3,4,5]
Output: 4
Explanation: Buy on day 1 (price = 1) and sell on day 5 (price = 5), profit = 5-1 = 4.
             Note that you cannot buy on day 1, buy on day 2 and sell them later, as you are
             engaging multiple transactions at the same time. You must sell before buying again.

Example 3:

Input: [7,6,4,3,1]
Output: 0
Explanation: In this case, no transaction is done, i.e. max profit = 0.


 */
public class LC_0123_BestTimetoBuyandSellStock_III {
    /*
    https://www.programcreek.com/2014/02/leetcode-best-time-to-buy-and-sell-stock-iii-java/
    Comparing to I and II, III limits the number of transactions to 2. This can be solve by "devide and conquer".
    We use left[i] to track the maximum profit for transactions before i, and use right[i] to track the maximum profit
    for transactions after i.

    https://www.youtube.com/watch?v=LX8i5PxRVCg
     */
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }

        //max profit in 0 ... i
        int[] left = new int[prices.length];
        //max profit from  i ... price.length -1
        int[] right = new int[prices.length];

        // DP from left to right
        //left[0] = 0;
        int minPrice = prices[0];
        for (int i = 1; i < prices.length; i++) {
            minPrice = Math.min(minPrice, prices[i]);
            left[i] = Math.max(left[i - 1], prices[i] - minPrice);
        }

        // DP from right to left
        //right[prices.length - 1] = 0;
        int maxPrice = prices[prices.length - 1];
        for (int i = prices.length - 2; i >= 0; i--) {
            maxPrice = Math.max(maxPrice, prices[i]);
            right[i] = Math.max(right[i + 1], maxPrice - prices[i]);
        }

        int maxProfit = 0;
        for (int i = 0; i < prices.length; i++) {
            maxProfit = Math.max(maxProfit, left[i] + right[i]);
        }

        return maxProfit;
    }
}

File: 0123_Best_Time_to_Buy_and_Sell_Stock_III/solution.cpp
/*
123. Best Time to Buy and Sell Stock III

https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/
*/
#include <vector>

using namespace std;
/*
    https://www.programcreek.com/2014/02/leetcode-best-time-to-buy-and-sell-stock-iii-java/
    Comparing to I and II, III limits the number of transactions to 2. This can be solve by "devide and conquer".
    We use left[i] to track the maximum profit for transactions before i, and use right[i] to track the maximum profit
    for transactions after i.

    https://www.youtube.com/watch?v=LX8i5PxRVCg
     */
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        if (prices.empty())
        {
            return 0;
        }

        //max profit in 0 ... i
        vector<int> left(prices.size());
        //max profit from  i ... price.length -1
        vector<int> right(prices.size());

        // DP from left to right
        //left[0] = 0;
        int minPrice = prices[0];
        for (int i = 1; i < prices.size(); i++)
        {
            minPrice = min(minPrice, prices[i]);
            left[i] = max(left[i - 1], prices[i] - minPrice);
        }

        // DP from right to left
        //right[prices.length - 1] = 0;
        int maxPrice = prices[prices.size() - 1];
        for (int i = prices.size() - 2; i >= 0; i--)
        {
            maxPrice = max(maxPrice, prices[i]);
            right[i] = max(right[i + 1], maxPrice - prices[i]);
        }

        int maxProfit = 0;
        for (int i = 0; i < prices.size(); i++)
        {
            maxProfit = max(maxProfit, left[i] + right[i]);
        }

        return maxProfit;
    }
};